### ISSUE TITLE
how are the batch norm layers being updated during training process

### ISSUE THREAD

**Issue by sbashett** [automan000/Camera-based-Person-ReID#8]
Hello, I greatly appreciate the work and its pretty resourceful. 
I am trying to understand the training process and my question here is how are the running means and variances of batch norm layers going to get affected over time during the training process if I have data from multiple cameras. 

Say, if I am training on a single gpu, I would have only one replica created in your custom DataParallel module and all the sub-batches belonging to different cameras would be sent through the same replica or model. Wouldn't this cause the running means and variances of batch norm layers to be updated using data from all the cams? I understand that we would eventually drop these values during testing phase but this would cause the batch norm updates to be uneven during training phase and the network wouldn't learn efficiently. Moreover, wouldn't it defeat the purpose of isolating the batch norm values based on camera specific information? Please share your thoughts on this. Thanks in advance.

**Comment by automan000** [comment]
Hi, Thanks for your question.
If I understand correctly, the question is how to normalization each camera separately using only one replica.
There are two kinds of statistics inside a PyTorch BN layer: (1) **batch mean/var and** (2) **running mean/var**.

The BN layer does not use running mean/var in the training stage.
The batch mean/var is the one used during training.
They are calculated from each mini-batch **independently** and used to normalize the corresponding data.
We take advantage of this fact and achieve independent normalization for all training cameras.

The running mean/var is the one used in the testing phase.
They are accumulated from all mini-batch mean/var observed during training.
As you mentioned, they are incorrect in our framework and we will eventually replace them.

You can find more details at https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py

FYI, the hyper-parameter **momentum** does not affect batch mean/var. It is meaningless in our training stage.

**Comment by sbashett** [comment]
Thanks for the reply, that makes things a lot clear. May I ask which parts are you trying to replace in the framework and is this going to be a quick fix or would happen much later?

**Comment by automan000** [comment]
As described in our paper, we claim that **_the ReID data should be organized in a camera-wise manner rather than the conventional dataset-wise manner_**. 

- In the training stage, our custom [CamDataParallel](https://github.com/automan000/Camera-based-Person-ReID/blob/master/frameworks/training/data_parallel.py#L9) ensures that each training mini-batch contains images from the same camera.
- In the testing stage, [this procedure](https://github.com/automan000/Camera-based-Person-ReID/blob/master/test_model.py#L51) estimates the BN statistics for each camera and generates the correct features for all related images.

All these modifications are about organizing and processing data in a camera-wise manner.
They are **transparent** to the network architecture, objective functions, and training strategy.
In theory, all previous state-of-the-art methods can be integrated with our method.
More details can be found in our readme.

**Comment by sbashett** [comment]
got it, thanks for the information. I am closing this issue.
